File: gva/web/src/api/pmocker/deliverable.js
import service from '@/utils/request'

// @Summary 创建交付物
// @Router /deliverable/create [post]
export const createDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/create', method: 'post', data })
}

// @Summary 删除交付物
// @Router /deliverable/delete [delete]
export const deleteDeliverable = (params) => {
  return service({ url: '/pmocker/deliverable/delete', method: 'delete', params })
}

// @Summary 更新交付物
// @Router /deliverable/update [put]
export const updateDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/update', method: 'put', data })
}

// @Summary 查询交付物详情
// @Router /deliverable/find [get]
export const findDeliverable = (params) => {
  return service({ url: '/pmocker/deliverable/find', method: 'get', params })
}

// @Summary 获取交付物列表
// @Router /deliverable/list [get]
export const getDeliverableList = (params) => {
  return service({ url: '/pmocker/deliverable/list', method: 'get', params })
}

// @Summary 提交交付物评审
// @Router /deliverable/submitReview [post]
export const submitDeliverableReview = (data) => {
  return service({ url: '/pmocker/deliverable/submitReview', method: 'post', data })
}

// @Summary 接收交付物
// @Router /deliverable/accept [post]
export const acceptDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/accept', method: 'post', data })
}

// @Summary 驳回交付物
// @Router /deliverable/reject [post]
export const rejectDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/reject', method: 'post', data })
}

// @Summary 创建新版本
// @Router /deliverable/createVersion [post]
export const createDeliverableVersion = (data) => {
  return service({ url: '/pmocker/deliverable/createVersion', method: 'post', data })
}

// @Summary 创建基线
// @Router /deliverable/baseline [post]
export const createDeliverableBaseline = (data) => {
  return service({ url: '/pmocker/deliverable/baseline', method: 'post', data })
}

// @Summary 检出交付物（排他锁定）
// @Router /deliverable/checkOut [post]
export const checkOutDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/checkOut', method: 'post', data })
}

// @Summary 检入交付物（解锁并可选记录版本）
// @Router /deliverable/checkIn [post]
export const checkInDeliverable = (data) => {
  return service({ url: '/pmocker/deliverable/checkIn', method: 'post', data })
}

// @Summary 获取版本列表
// @Router /deliverable/listVersions [get]
export const getDeliverableVersions = (params) => {
  return service({ url: '/pmocker/deliverable/listVersions', method: 'get', params })
}

// @Summary 获取追溯报告
// @Router /deliverable/traceReport [get]
export const getDeliverableTraceReport = (params) => {
  return service({ url: '/pmocker/deliverable/traceReport', method: 'get', params })
}

// @Summary 获取交付物统计
// @Router /deliverable/stats [get]
export const getDeliverableStats = (params) => {
  return service({ url: '/pmocker/deliverable/stats', method: 'get', params })
}

File: gva/server/plugin/pmocker_deliverable/api/deliverable.go
package deliverable

import (
	"strconv"

	"github.com/flipped-aurora/gin-vue-admin/server/model/common/response"
	"github.com/flipped-aurora/gin-vue-admin/server/utils"
	eavtypes "github.com/pmocker-io/pmocker/pkg/pmocker/eav"
	"github.com/gin-gonic/gin"
)

type Api struct{}

func (a *Api) Create(c *gin.Context) {
	var req struct {
		ProjectID uint                   `json:"projectId"`
		Title     string                 `json:"title"`
		Attrs     map[string]interface{} `json:"attrs"`
		CreatorID uint                   `json:"creatorId"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	id, err := ServiceGroupApp.CreateDeliverable(c.Request.Context(), req.ProjectID, req.Title, req.Attrs, req.CreatorID)
	if err != nil {
		response.FailWithMessage("创建失败: "+err.Error(), c)
		return
	}
	response.OkWithData(gin.H{"id": id}, c)
}

func (a *Api) Delete(c *gin.Context) {
	id, err := strconv.ParseUint(c.Query("id"), 10, 64)
	if err != nil {
		response.FailWithMessage("参数错误: id 必填", c)
		return
	}
	if err := ServiceGroupApp.DeleteDeliverable(c.Request.Context(), uint(id)); err != nil {
		response.FailWithMessage("删除失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

func (a *Api) Update(c *gin.Context) {
	var e eavtypes.Entity
	if err := c.ShouldBindJSON(&e); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if err := ServiceGroupApp.UpdateDeliverable(c.Request.Context(), e, utils.GetUserID(c)); err != nil {
		response.FailWithMessage("更新失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("更新成功", c)
}

// CheckOut 检出交付物（排他锁定）
func (a *Api) CheckOut(c *gin.Context) {
	var req struct {
		ID uint `json:"id"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if req.ID == 0 {
		response.FailWithMessage("参数错误: id 必填", c)
		return
	}
	if err := ServiceGroupApp.CheckOut(c.Request.Context(), req.ID, utils.GetUserID(c)); err != nil {
		response.FailWithMessage("检出失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("检出成功", c)
}

// CheckIn 检入交付物（解锁并可选记录版本）
func (a *Api) CheckIn(c *gin.Context) {
	var req struct {
		ID          uint   `json:"id"`
		VersionNote string `json:"versionNote"`
		FileRef     string `json:"fileRef"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if req.ID == 0 {
		response.FailWithMessage("参数错误: id 必填", c)
		return
	}
	if err := ServiceGroupApp.CheckIn(c.Request.Context(), req.ID, utils.GetUserID(c), req.VersionNote, req.FileRef, utils.GetUserAuthorityId(c)); err != nil {
		response.FailWithMessage("检入失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("检入成功", c)
}

func (a *Api) Find(c *gin.Context) {
	id, err := strconv.ParseUint(c.Query("id"), 10, 64)
	if err != nil {
		response.FailWithMessage("参数错误: id 必填", c)
		return
	}
	e, err := ServiceGroupApp.GetDeliverable(c.Request.Context(), uint(id))
	if err != nil {
		response.FailWithMessage("查询失败: "+err.Error(), c)
		return
	}
	response.OkWithData(e, c)
}

func (a *Api) List(c *gin.Context) {
	projectID, _ := strconv.ParseUint(c.Query("projectId"), 10, 64)
	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))
	list, total, err := ServiceGroupApp.ListDeliverables(c.Request.Context(), uint(projectID), offset, limit)
	if err != nil {
		response.FailWithMessage("查询失败: "+err.Error(), c)
		return
	}
	response.OkWithDetailed(gin.H{"list": list, "total": total}, "查询成功", c)
}

func (a *Api) SubmitReview(c *gin.Context) {
	var req struct {
		ID uint `json:"id"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if err := ServiceGroupApp.SubmitForReview(c.Request.Context(), req.ID); err != nil {
		response.FailWithMessage("提交评审失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("已提交评审", c)
}

func (a *Api) Accept(c *gin.Context) {
	var req struct {
		ID              uint   `json:"id"`
		ReviewerComment string `json:"reviewerComment"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if err := ServiceGroupApp.AcceptDeliverable(c.Request.Context(), req.ID, req.ReviewerComment); err != nil {
		response.FailWithMessage("验收失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("验收通过", c)
}

func (a *Api) Reject(c *gin.Context) {
	var req struct {
		ID     uint   `json:"id"`
		Reason string `json:"reason"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if err := ServiceGroupApp.RejectDeliverable(c.Request.Context(), req.ID, req.Reason); err != nil {
		response.FailWithMessage("驳回失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("已驳回", c)
}

func (a *Api) CreateVersion(c *gin.Context) {
	var req struct {
		DeliverableID uint   `json:"deliverableId"`
		Version       string `json:"version"`
		ChangeLog     string `json:"changeLog"`
		CreatorID     uint   `json:"creatorId"`
		FileRef       string `json:"fileRef"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	result, err := ServiceGroupApp.CreateVersion(c.Request.Context(), req.DeliverableID, req.Version, req.ChangeLog, req.CreatorID, req.FileRef)
	if err != nil {
		response.FailWithMessage("创建版本失败: "+err.Error(), c)
		return
	}
	response.OkWithData(result, c)
}

func (a *Api) ListVersions(c *gin.Context) {
	deliverableID, _ := strconv.ParseUint(c.DefaultQuery("deliverableId", "0"), 10, 64)
	list, err := ServiceGroupApp.ListVersions(c.Request.Context(), uint(deliverableID))
	if err != nil {
		response.FailWithMessage("查询失败: "+err.Error(), c)
		return
	}
	response.OkWithData(gin.H{"list": list}, c)
}

func (a *Api) Baseline(c *gin.Context) {
	var req struct {
		ID uint `json:"id"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.FailWithMessage("参数错误: "+err.Error(), c)
		return
	}
	if err := ServiceGroupApp.BaselineDeliverable(c.Request.Context(), req.ID); err != nil {
		response.FailWithMessage("基线化失败: "+err.Error(), c)
		return
	}
	response.OkWithMessage("已基线化", c)
}

func (a *Api) TraceReport(c *gin.Context) {
	projectID, _ := strconv.ParseUint(c.Query("projectId"), 10, 64)
	report, err := ServiceGroupApp.TraceabilityReport(c.Request.Context(), uint(projectID))
	if err != nil {
		response.FailWithMessage("生成追溯报告失败: "+err.Error(), c)
		return
	}
	response.OkWithData(report, c)
}

func (a *Api) Stats(c *gin.Context) {
	projectID, _ := strconv.ParseUint(c.Query("projectId"), 10, 64)
	stats, err := ServiceGroupApp.DeliverableStats(c.Request.Context(), uint(projectID))
	if err != nil {
		response.FailWithMessage("统计失败: "+err.Error(), c)
		return
	}
	response.OkWithData(stats, c)
}
